src/flow/src/utils.rs

```diff
@@ -0,0 +1,465 @@
+    pub fn set_compaction(&mut self, now: Timestamp) -> Result<(), EvalError> {
+        let mut should_compact = self.spine.split_off(&(now + 1));
+        std::mem::swap(&mut should_compact, &mut self.spine);
+
+        // if a full arrangement is not needed, we can just discard everything before and including now
+        if !self.full_arrangement {
+            return Ok(());
+        }
+        // else we update them into current key value pairs
+        let mut compacted_batch: BTreeMap<Row, SmallVec<[DiffRow; 2]>> = Default::default();
+
+        for (_, batch) in should_compact {
+            for (key, updates) in batch {
+                if let Some(s) = &mut self.expire_state {
+                    if s.is_key_expired(now, &key)? {
+                        continue;
+                    }
+                    s.update_event_ts(now, &key)?;
+                }
+                // if diff cancel out each other, then remove the key
+                let mut old_row: Option<DiffRow> =
+                    compacted_batch.get(&key).and_then(|v| v.first()).cloned();
+
+                for new_row in updates {
+                    old_row = compact_diff_row(old_row, new_row);
+                }
+                if let Some(compacted_update) = old_row {
+                    compacted_batch.insert(key, smallvec![compacted_update]);
+                } else {
+                    compacted_batch.remove(&key);
+                }
+            }
+        }
+
+        // insert the compacted batch into spine with key being `now`
+        let mut first_entry = BTreeMap::from([(now, compacted_batch)]);
+        self.spine.append(&mut first_entry);
+        Ok(())
+    }
+
+    /// get the updates of the arrangement from the given range of time
+    pub fn get_updates_in_range<R: std::ops::RangeBounds<Timestamp>>(
+        &self,
+        range: R,
+    ) -> Vec<KeyValDiffRow> {
+        let mut result = vec![];
+        for (_ts, batch) in self.spine.range(range) {
+            for (key, updates) in batch.clone() {
+                for (val, ts, diff) in updates {
+                    result.push(((key.clone(), val), ts, diff));
+                }
+            }
+        }
+        result
+    }
+
+    /// expire keys in now that are older than expire_time, intended for reducing memory usage and limit late data arrive
+    pub fn trunc_expired(&mut self, now: Timestamp) {
+        if let Some(s) = &mut self.expire_state {
+            let expire_time = if let Some(t) = s.get_expire_time(now) {
+                t
+            } else {
+                // never expire
+                return;
+            };
+            // find all keys smaller than or equal expire_time and silently remove them
+            let mut after = s.event_ts_to_key.split_off(&(expire_time + 1));
+            std::mem::swap(&mut s.event_ts_to_key, &mut after);
+            let before = after;
```

Comment:
It feels like a bit of abstract leakage

Reply:
> It feels like a bit of abstract leakage

If you are refering to using `drain_range`, sadly there are cavets related to lifetime and soundness, so no avaliable functions now: https://internals.rust-lang.org/t/pre-rfc-btreemap-btreeset-drain-with-a-range/15911 maybe I can wrap `split_off` into a simple handwritten drain_range that `drain_range(input_map)->(drain_result_map, drain_rest_of_map)`? On a second thought, that might be a good idea